crypto/hpke/hpke.c

```diff
@@ -0,0 +1,2366 @@
+            ERR_raise(ERR_LIB_CRYPTO, ERR_R_INTERNAL_ERROR);
+            goto err;
+        }
+        if (EVP_PKEY_fromdata(ctx, &lpriv, EVP_PKEY_KEYPAIR, params) <= 0) {
+            ERR_raise(ERR_LIB_CRYPTO, ERR_R_INTERNAL_ERROR);
+            goto err;
+        }
+    }
+    if (lpriv == NULL) {
+        /* check PEM decode - that might work :-) */
+        BIO *bfp = BIO_new(BIO_s_mem());
+
+        if (bfp == NULL) {
+            ERR_raise(ERR_LIB_CRYPTO, ERR_R_INTERNAL_ERROR);
+            goto err;
+        }
+        BIO_write(bfp, prbuf, prbuf_len);
+        if (!PEM_read_bio_PrivateKey(bfp, &lpriv, NULL, NULL)) {
+            BIO_free_all(bfp);
+            bfp = NULL;
+            ERR_raise(ERR_LIB_CRYPTO, ERR_R_INTERNAL_ERROR);
+            goto err;
+        }
+        if (bfp != NULL) {
+            BIO_free_all(bfp);
+            bfp = NULL;
+        }
+        if (lpriv == NULL) {
+            /* if not done, prepend/append PEM header/footer and try again */
+            memcpy(hf_prbuf, PEM_PRIVATEHEADER, strlen(PEM_PRIVATEHEADER));
+            hf_prbuf_len += strlen(PEM_PRIVATEHEADER);
+            memcpy(hf_prbuf + hf_prbuf_len, prbuf, prbuf_len);
+            hf_prbuf_len += prbuf_len;
+            memcpy(hf_prbuf + hf_prbuf_len, PEM_PRIVATEFOOTER,
+                   strlen(PEM_PRIVATEFOOTER));
+            hf_prbuf_len += strlen(PEM_PRIVATEFOOTER);
+            bfp = BIO_new(BIO_s_mem());
+            if (bfp == NULL) {
+                ERR_raise(ERR_LIB_CRYPTO, ERR_R_INTERNAL_ERROR);
+                goto err;
+            }
+            BIO_write(bfp, hf_prbuf, hf_prbuf_len);
+            if (!PEM_read_bio_PrivateKey(bfp, &lpriv, NULL, NULL)) {
+                BIO_free_all(bfp);
+                bfp = NULL;
+                ERR_raise(ERR_LIB_CRYPTO, ERR_R_INTERNAL_ERROR);
+                goto err;
+            }
+            BIO_free_all(bfp);
+            bfp = NULL;
+        }
+    }
+    if (lpriv == NULL) {
+        ERR_raise(ERR_LIB_CRYPTO, ERR_R_INTERNAL_ERROR);
+        goto err;
+    }
+    *retpriv = lpriv;
+    erv = 1;
+
+err:
+#ifndef OPENSSL_NO_EC
+    BN_free(calc_priv);
+    EC_POINT_free(calc_pub);
+    EC_GROUP_free(curve);
+#endif
+    BN_free(priv);
+    EVP_PKEY_CTX_free(ctx);
+    OSSL_PARAM_BLD_free(param_bld);
+    OSSL_PARAM_free(params);
+    return erv;
+}
+
+/**
+ * @brief check if a suite is supported locally
+ *
+ * @param suite is the suite to check
+ * @return 1 for good/supported, not 1 otherwise
+ */
+static int hpke_suite_check(OSSL_HPKE_SUITE suite)
+{
+    /*
+     * Check that the fields of the suite are each
+     * implemented here
+     */
+    int kem_ok = 0;
+    int kdf_ok = 0;
+    int aead_ok = 0;
+    int ind = 0;
+    int nkems = OSSL_NELEM(hpke_kem_tab);
+    int nkdfs = OSSL_NELEM(hpke_kdf_tab);
+    int naeads = OSSL_NELEM(hpke_aead_tab);
+
+    /* check KEM */
+    for (ind = 0; ind != nkems; ind++) {
+        if (suite.kem_id == hpke_kem_tab[ind].kem_id) {
```

Comment:
I would suggest getting rid of the empty entry in your tables.. This currently passes with 0.